src/main/java/com/mongodb/hibernate/jdbc/MongoConnection.java

```diff
@@ -56,43 +59,67 @@ final class MongoConnection extends ConnectionAdapter {
     @Override
     public void setAutoCommit(boolean autoCommit) throws SQLException {
         checkClosed();
-        throw new NotYetImplementedException(
-                "To be implemented in scope of https://jira.mongodb.org/browse/HIBERNATE-30");
+        if (autoCommit != this.autoCommit) {
+            try {
+                if (clientSession.hasActiveTransaction()) {
+                    clientSession.commitTransaction();
+                }
+            } catch (RuntimeException e) {
+                throw new SQLException("Failed to commit transaction", e);
+            }
+            try {
+                if (!autoCommit) {
+                    clientSession.startTransaction();
+                }
+                this.autoCommit = autoCommit;
+            } catch (RuntimeException e) {
+                throw new SQLException("Failed to start transaction", e);
+            }
```

Comment:
that is a good point. The concern here is Hibernte ORM has dictated the transaction pattern in a framework way so we don't have luxurary to go about customization as PostgreSQL. Let me explore its code logic to see whether there is some way out.

From my current understanding, Hibernate ORM (it has a `transaction` root package in its codebase), it supports two ways: 1) JTA; 2) resource local or JDBC driver way (as per point No.2 above). It doesn't support deep customization for JTA mode, but that is out of question for our JDBC emulation approach.

We can discuss later after I read through the relevant Hibernate code logic.

Reply:
JDBC has no explicit transaction beginning statement, and yeah, `setAutoCommit(false)` is the precondition to start a transaction implicitly (it is up to JDBC Driver to do some optimization as PostgreSQL). We are emulating JDBC flow using Mongo Java driver and live in the tight jacket of Hibernate ORM framework. Maybe it is a good perf opt ticket in the future (if it is really worthwhile), but for now I guess we have limited choice here.

Reply:
So here is the relevant Hibernate User Guide: https://docs.jboss.org/hibernate/orm/6.6/userguide/html_single/Hibernate_User_Guide.html#transactions-api
<img width="940" alt="Screenshot 2024-12-05 at 2 23 55 PM" src="https://github.com/user-attachments/assets/c04d7702-460b-4d51-aa17-97221997690d">

The relevant code logic is at https://github.com/hibernate/hibernate-orm/blob/main/hibernate-core/src/main/java/org/hibernate/resource/jdbc/internal/AbstractLogicalConnectionImplementor.java#L69

So it seems Hibernate transaction demarcation is purely based on `setAutoCommit(false)`. Particularly, the transaction API will set its state based on the assumption that once `setAutoCommit(false)` is called, the transaction state will be `ACTIVE`. For smooth Hibernate integration, we'd better align with Hibernate's expectation.

So the only question outstanding is:

- could we delay `clientSession.startTransaction()` lazily when the first `java.sql.Statement` invocation is called?

Well, it seems doable (if so, we could hold off the review of this PR until we finished the whole JDBC flow, that is fine), but it might bring about lots of complexity, including:

- Hibernate integration might be broken and we need to fix all kinds of edge case (like some exception is thrown from current session prior to first statement issuing or delayed transaction starting, then Hibernate will try to rollback its transaction and hopefully it won't trigger some of our internal SQLException or even worse, some AssertError)
- unit testing and integration testing will be challenging for we need to take into consideration all the lazy initialization corner cases

Another option is to take a simple approach without emulating PostgreSQL and conduct perf optimization after milestone No. 1. I am afraid if we implement the lazy demarcation idea, milestone will be postponed again and again.

How do you think?

Reply:
I think you are right. `setAutoCommit(false)` should not trigger transaction starting. It should be done lazily during `Statement` or `PreparedStatement` executing.
Updated the PR. Tons of thanks to your good catch!